src/devices/cpu/m6502/gew12.cpp

```diff
@@ -0,0 +1,199 @@
+
+	// TODO: UART
+	// map(0x1f11, 0x1f11) Rx data
+	// map(0x1f12, 0x1f12) status (bits 0-1: error, bit 2: Rx ready, bit 3: Tx ready)
+	// map(0x1f13, 0x1f13) Tx data
+
+	map(0x1f20, 0x1f23).w(FUNC(gew12_device::timer_count_w));
+	map(0x1f24, 0x1f25).r(FUNC(gew12_device::timer_count_r));
+
+	// TODO - only the last port (PL) is known for sure
+	// on psr160/260, the rest seem to all be handled by built-in key/button scanning
+	map(0x1f2a, 0x1f2f).rw(FUNC(gew12_device::port_r), FUNC(gew12_device::port_w));
+	map(0x1f3a, 0x1f3f).rw(FUNC(gew12_device::port_ddr_r), FUNC(gew12_device::port_ddr_w));
+
+	map(0x1f52, 0x1f52).lw8(NAME([this](uint8_t data) { m_bank[0]->set_entry(data); }));
+	map(0x1f53, 0x1f53).nopw(); // bank 0 MSB? (always zero)
+	map(0x1f54, 0x1f54).lw8(NAME([this](uint8_t data) { m_bank[1]->set_entry(data); }));
+	map(0x1f55, 0x1f55).nopw(); // bank 1 MSB? (always zero)
+
+	map(0x4000, 0x7fff).bankr(m_bank[0]);
+	map(0x8000, 0xbfff).bankr(m_bank[1]);
+	map(0xc000, 0xffff).rom().region(DEVICE_SELF, 0);
+}
+
+
+u8 gew12_device::irq_stat_r()
+{
+	return m_irq_pending & m_irq_enable;
+}
+
+void gew12_device::irq_en_w(u8 data)
+{
+	m_irq_enable = data;
+	irq_update();
+}
+
+void gew12_device::internal_irq(int num, int state)
+{
+	if (state == ASSERT_LINE)
+		m_irq_pending |= (1 << num);
+	else
+		m_irq_pending &= ~(1 << num);
+
+	irq_update();
+}
+
+void gew12_device::irq_update()
+{
+	if (m_irq_pending & m_irq_enable)
+		set_input_line(M65C02_IRQ_LINE, ASSERT_LINE);
+	else
+		set_input_line(M65C02_IRQ_LINE, CLEAR_LINE);
+}
+
+
+void gew12_device::timer_stat_w(u8 data)
+{
+	if (BIT(data, 0))
+		internal_irq(INTERNAL_IRQ_TIMER0, CLEAR_LINE);
+	if (BIT(data, 1))
+		internal_irq(INTERNAL_IRQ_TIMER1, CLEAR_LINE);
+}
+
+u8 gew12_device::timer_count_r(offs_t offset)
+{
+	// both timer IRQ handlers read the current count from the same registers
+	// apparently reading timer 1 count if its IRQ is pending, else timer 0 count
+	const unsigned timer = BIT(m_irq_pending, INTERNAL_IRQ_TIMER1);
+
+	if (!BIT(offset, 0))
+		return m_timer_count[timer];
+	else
+		return m_timer_count[timer] >> 8;
+}
+
+void gew12_device::timer_count_w(offs_t offset, u8 data)
+{
+	const unsigned timer = offset >> 1;
+
+	if (!BIT(offset, 0))
+		m_timer_count[timer] = (m_timer_count[timer] & 0xff00) | data;
+	else
+		m_timer_count[timer] = (m_timer_count[timer] & 0x00ff) | (data << 8);
+}
+
+TIMER_CALLBACK_MEMBER(gew12_device::timer_tick)
+{
+	if (!m_timer_count[param])
+		internal_irq(INTERNAL_IRQ_TIMER0 + param, ASSERT_LINE);
+
+	m_timer_count[param]--;
+}
+
+
+u8 gew12_device::port_r(offs_t offset)
+{
+	const u8 out_data = m_port_data[offset] & ~m_port_ddr[offset];
+	return out_data | (m_in_cb[offset]() & m_port_ddr[offset]);
+}
+
+void gew12_device::port_w(offs_t offset, u8 data)
+{
+	m_port_data[offset] = data;
+
+	const u8 out_data = m_port_data[offset] & ~m_port_ddr[offset];
+	m_out_cb[offset](out_data | (m_in_cb[offset]() & m_port_ddr[offset]));
+}
```

Comment:
Good call, it totally slipped my mind that callbacks could be called with the mask as a separate argument. That also seems to eliminate the need for the `m_ignore_in_ddr` hack, so that'll be removed also.

Reply:
The reason it's done like that currently is because all of the GEW7-based models do button matrix scanning via the DDR registers by setting one row to output and the rest to input, and expecting the ones set as inputs to remain pulled up externally (except on psr180 and psr76 where the matrix rows are active-high through a transistor array instead). There's probably a better way to handle that though.

Reply:
That should be handled externally – you just feed send what the chip is driving on the output pins.  You can get weird feedback loops when you try to mix the input state in like this.

For example see the code for the 68705 family here: https://github.com/mamedev/mame/blob/mame0253/src/devices/cpu/m6805/m68705.cpp#L299

If you’ve got open drain outputs and high bits in the DDR are inputs, you want to do something like:
```c++
	m_out_cb[offset](0, m_port_data[offset] | m_port_ddr[offset], ~(m_port_data[offset] | m_port_ddr[offset]));
```

If you’ve got tri-state outputs and high bits in the DDR are inputs, you want to do something like:
```c++
	m_out_cb[offset](0, m_port_data[offset], ~m_port_ddr[offset]);
```

The offset is unused and hence always zero.  The mask should tell the client which lines are being actively driven.

Reply:
Neither of those approaches seem to work; the resulting calls to `psr150_state::keys_w` behave like all of the bits are being driven simultaneously regardless of what's passed to the mask. Even ignoring the DDR and passing a hardcoded 0 as the mask produces the same behavior.

Reply:
**That should be handled externally.**  The code that receives the notification when the output is updated should deal with any external signals that also affect what it sees on the lines.  Making the CPU call the input callback like this on updating the outputs leads to weird feedback loops and other difficult-to-diagnose bugs.